### PR TITLE
Add initial tests

### PR DESCRIPTION
**Pull request by MaHermann**
This adds a base set of tests for the core functionalities. There is still a lot to be desired, especially more edge cases and thorough testing of `TreePartition` and the parallelization in `BoxMap`, but I hope it is at least a solid foundation for more tests and helpful in this phase of development.

### CODE REVIEW

File: src/box.jl [hunk 1]
```diff
@@ -18,3 +18,9 @@ struct Box{N,T}
         return new{N,T}(SVector{N,T}(center), SVector{N,T}(radius))
     end
 end
+
+
+function contains(box::Box{N,T}, point::SVector{N,T}) where {N,T}
```
Comment (cafaxo) [inline src/box.jl:23]: It does not seem correct to add this function in this PR

File: test/algorithms.jl [hunk 2]
```diff
@@ -0,0 +1,28 @@
+using GAIO, StaticArrays, Test
+
+@testset "exported functionality" begin
+    # a toy example with an easy attractor to test how well all the components
+    # fit together
+    f(x::SVector{2,Float64}) = SVector(x[1],x[2]*0.5)
+    test_points = [SVector(-1.,-1.),SVector(-1.,1.),
```
Comment (cafaxo) [inline test/algorithms.jl:7]: Please format this properly. (e.g. space before commas, except in type parameters such as `SVector{2,Float64}`)

File: test/algorithms.jl [hunk 3]
```diff
@@ -0,0 +1,28 @@
+using GAIO, StaticArrays, Test
+
+@testset "exported functionality" begin
+    # a toy example with an easy attractor to test how well all the components
+    # fit together
+    f(x::SVector{2,Float64}) = SVector(x[1],x[2]*0.5)
```
Comment (cafaxo) [inline test/algorithms.jl:6]: No need to specify the type of `x` here.

File: test/algorithms.jl [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+                   SVector(1.,-1.),SVector(1.,1.)]
+    g = boxmap(f,test_points)
+
+    n = 10
+    domain = Box(SVector(0.,0.),SVector(1.,1.))
+    partition = RegularPartition(domain)
+    partition_at_depth_n = RegularPartition(domain,n)
+    rga = relative_attractor(partition[:],g,n)
+    unstable = unstable_set!(partition_at_depth_n[:],g)
+    #ground truths attractor and unstable set
```
Comment (cafaxo) [inline test/algorithms.jl:17]: Space after #

File: test/boxmap.jl [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    # easiest way right now to check for equality
+    @test Base.length(union(image,mapped1)) == Base.length(intersect(image,mapped1))
+    @test Base.length(union(image,mapped2)) != Base.length(intersect(image,mapped2))
+    @test Base.length(union(image,mapped3)) == Base.length(intersect(image,mapped4))
+    @test Base.length(union(image,mapped4)) == Base.length(intersect(image,mapped3))
+    # check for subset
+    @test Base.length(intersect(boxset,mapped2)) == Base.length(mapped2)
+
+
+
```
Comment (cafaxo) [inline test/boxmap.jl:47]: please only one newline at most

File: test/box.jl [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @test !GAIO.contains(box,on_boundary_right)
+    @test !GAIO.contains(box,outside_left)
+    @test !GAIO.contains(box,outside_right)
+
+
+    center = SVector(0.0,0.0,0.0)
+    radius = SVector(1.0,1.0)
+    @test_throws Exception Box(center,radius)
+
+    center = SVector(0.0,0.0)
```
Comment (cafaxo) [inline test/box.jl:42]: Please split stuff like this up into mutiple testsets (e.g. "negative radius") and try to not shadow variables.

File: test/boxset.jl [hunk 7]
```diff
@@ -0,0 +1,140 @@
+using GAIO, StaticArrays, Test
+
+@testset "exported functionality" begin
+    @testset "regular partition" begin
```
Comment (cafaxo) [inline test/boxset.jl:4]: Also needs multiple testsets; each with a name describing what you are testing.

File: test/runtests.jl [hunk 8]
```diff
@@ -1,6 +1,10 @@
-using GAIO
-using Test
+using GAIO, SafeTestsets
 
-@testset "GAIO.jl" begin
-    # TODO (include files here)
+@safetestset "GAIO.jl" begin
+    @safetestset "Box" begin include("box.jl") end
```
Comment (cafaxo) [inline test/runtests.jl:4]: Why do we need safetestset?

Comment (MaHermann) [reply test/runtests.jl:4]: As far as I understood, safetestset puts every testset into its own module, so a variable like `box` can't leak out of one of them into another. In such a case, potentially the order or execution could matter etc.

Comment (cafaxo) [reply test/runtests.jl:4]: No; variables do not leak out of begin end blocks. We do not need safetestset for this.

File: src/box.jl (hunk 1 above)
Comment (gaioguy) [reply src/box.jl:23]: Yes, this is useful, but should be in a separate PR, maybe together with other misc small additions.

Comment (MaHermann) [reply src/box.jl:23]: It makes the tests a bit cleaner at some points, but for now I'll just put it in some "testutils" or the like.

As you said, we can add it to the main module later.

Comment (cafaxo) [reply src/box.jl:23]: The correct name for this is "Base.in". With this, we have the syntax `point in box`. I opened a PR at #27.

**Comment by cafaxo** [comment]
Concerning the formatting:
There is this: https://github.com/domluna/JuliaFormatter.jl
and that: https://github.com/julia-actions/julia-format
Should we add this as a Github Action:
"Github action which verifies JuliaFormatter.jl was on .jl files."?

**Comment by MaHermann** [comment]
Looks good! Just to make 100% sure I understand this correctly: The Github Action would not automatically format the code (in the way e.g. pre-commit hooks can do) but only check for compliance, right? So we still need to locally run the formatter (which of course is no big deal, I just want to be sure)

**Comment by cafaxo** [comment]
Yes, you would still locally need to run it or be compliant :)
I'll look into it. If it is sufficiently configurable, I will open a PR that adds the action.

**Comment by cafaxo** [comment]
After trying out JuliaFormatter, I do not think that we should install this action. The reason is that JuliaFormatter messes up manually aligned lines.
Other than that, JuliaFormatter seems nice (especially with [this](https://domluna.github.io/JuliaFormatter.jl/stable/yas_style/#Configuration-File-Example-1) configuration. So I will start using it locally.

**Comment by MaHermann** [comment]
I rebased this on #27, which admittedly might not be best practice, but significantly simpliefies things in this case. Still, this means we shouldn't merge this before the other one.

**Comment by cafaxo** [comment]
I just merged this, but Github did something really stupid:

> The rebase and merge behavior on GitHub deviates slightly from git rebase. Rebase and merge on GitHub will always update the committer information and create new commit SHAs, whereas git rebase outside of GitHub does not change the committer information when the rebase happens on top of an ancestor commit. For more information about git rebase, see the "Git rebase" chapter from the Pro Git book.
quoted from: https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/about-pull-request-merges#rebase-and-merge-your-pull-request-commits

Apparently, the webinterface does `git rebase --no-ff" which always creates a new commit hash. This would have forced you to rebase it AGAIN on master. To circumvent this nonsense, I just force pushed the correct commit from the PR to master. For the future, I will merge future PRs from the command line using fast-forward to prevent Github from generating useless new commit hashes.

File: test/box.jl [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        #boxes are half open to the right
+        @test left ∈ box
+        @test right ∉ box
+        @test on_boundary_left ∈ box
+        @test on_boundary_right ∉ box
+        @test outside_left ∉ box
+        @test outside_right ∉ box
+    end
+
+    @testset "non matching dimensions" begin
```
Comment (cafaxo) [inline test/box.jl:41]: Please be consistent with newlines before testset. :)
There are no newlines above, but here there are.

File: test/boxmap.jl [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        @test Base.length(intersect(boxset, mapped2)) == Base.length(mapped2)
+    end
+    @testset "points in BoxMaps" begin
+        f(x) = x .^ 2
+        test_points = [(-1.0, -1.0), (-1.0, 1.0), (1.0, -1.0), (1.0, 1.0)]
+        g = boxmap(f, test_points)
+        x = (-2.0, 3.0)
+        y = SVector(4.0, 1)
+        @test f(x) == (4.0, 9.0)    #sanity checks
+        @test f(y) == [16.0, 1.0]
```
Comment (cafaxo) [inline test/boxmap.jl:53]: What are you testing here?

File: test/algorithms.jl [hunk 11]
```diff
@@ -0,0 +1,32 @@
+using GAIO, StaticArrays, Test
```
Comment (cafaxo) [inline test/algorithms.jl:1]: Please one import per line.

File: test/boxset.jl [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        end
+        # this seems to be the easiest way right now
+        # to get a full tree at a certain depth. Not
+        # really very clean
+        partition = full_boxset.partition
+        p1 = SVector(0.5, 0.5, 0.5)
+        p2 = SVector(-0.5, 0.5, 0.5)
+        p3 = SVector(0.5, -0.5, 0.5)
+        p4 = SVector(0.51, 0.51, 0.51)
+        @testset "box sets created on points" begin
```
Comment (cafaxo) [inline test/boxset.jl:104]: newline before testset here, please
maybe always one newline before testset?

File: test/partition_regular.jl [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        @test size(GAIO.keys_all(partition)) == (2^depth(partition),)
+    end
+    inside = SVector(0.5, 0.5, 0.5)
+    left = SVector(-1.0, -1.0, -1.0)
+    right = SVector(1.0, 1.0, 1.0)
+    on_boundary_left = SVector(0.0, 0.0, -1.0)
+    on_boundary_right = SVector(0.0, 1.0, 0.0)
+    outside_left = SVector(0.0, 0.0, -2.0)
+    outside_right = SVector(0.0, 2.0, 0.0)
+    @testset "point to key" begin
```
Comment (cafaxo) [inline test/partition_regular.jl:50]: same thing as above

File: test/partition_tree.jl [hunk 14]
```diff
@@ -0,0 +1,26 @@
+using GAIO, StaticArrays, Test
+
+# this is still very basic functionality only
```
Comment (cafaxo) [inline test/partition_tree.jl:3]: remove this comment. we can always expand the tests later.
but is this the right place to observe that?

File: test/box.jl [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+@testset "internal functionality" begin
+    box = Box(SVector(0.0, 0.0), SVector(1.0, 1.0))
+    @testset "integer point in box" begin
+        point_int_outside = SVector(2, 2)
+        point_int_inside = SVector(0, 0)
+        @test point_int_inside ∈ box
+        @test point_int_outside ∉ box
+    end
+    @testset "point in box error" begin
+        point_3D = SVector(0.0, 0.0, 0.0)
```
Comment (cafaxo) [inline test/box.jl:63]: just call this `point`, or even simpler:
`@test_throws DimensionMismatch SVector(0.0, 0.0, 0.0) ∈ box`

File: test/boxset.jl [hunk 16]
```diff
@@ -0,0 +1,147 @@
+using GAIO, StaticArrays, Test
+
+@testset "exported functionality" begin
+    @testset "regular partition" begin
+        partition = RegularPartition(Box(SVector(0.0, 0.0, 0.0), SVector(1.0, 1.0, 1.0)),
```
Comment (cafaxo) [inline test/boxset.jl:5]: Seems that Julia Formatter messed this up...

File: test/boxset.jl [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+        p4 = SVector(0.51, 0.51, 0.51)
+        @testset "boxsets created on points" begin
+            box_set = partition[(p1, p2, p3, p4)]
+            @test Base.length(box_set) == 3
+            @test any(box -> p1 ∈ box, box_set)
+            @test any(box -> p2 ∈ box, box_set)
+            @test any(box -> p3 ∈ box, box_set)
+            @test any(box -> p4 ∈ box, box_set)
+            n = 10
+            for _ = 1:n
```
Comment (cafaxo) [inline test/boxset.jl:37]: By the way, the YAS Style (https://domluna.github.io/JuliaFormatter.jl/stable/yas_style/) recommends to put "in" instead of "=" here.

File: test/partition_regular.jl [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+    @testset "roundtrip" begin
+        point = SVector(0.3, 0.3, 0.3)
+        key = GAIO.point_to_key(partition, point)
+        box = GAIO.key_to_box(partition, key)
+        @test point ∈ box
+        key_2 = GAIO.point_to_key(partition, box.center)
+        @test key == key_2
+    end
+    @testset "points with wrong dimension" begin
+        point_2D = SVector(0.0, 0.0)
```
Comment (cafaxo) [inline test/partition_regular.jl:76]: local variables are always lowercase, i.e. `point_2d`

File: test/partition_regular.jl [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        for _ = 1:n
+            partition = subdivide(partition)
+        end
+        @test depth(partition) == n
+        @test dimension(partition) == 4
+    end
+    @testset "size" begin
+        partition = RegularPartition(Box(SVector(0.0, 1.0), SVector(1.0, 1.0)), 3)
+        @test depth(partition) == 3
+        @test Base.size(partition) == (4, 2)
```
Comment (cafaxo) [inline test/partition_regular.jl:21]: I don't think you need the Base. prefix here

File: test/partition_regular.jl [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+    @testset "key to point" begin
+        @test typeof(GAIO.key_to_box(partition, key_inside)) <: typeof(partition.domain)
+        @test GAIO.key_to_box(partition, key_inside) != GAIO.key_to_box(partition, key_left)
+    end
+    @testset "roundtrip" begin
+        point = SVector(0.3, 0.3, 0.3)
+        key = GAIO.point_to_key(partition, point)
+        box = GAIO.key_to_box(partition, key)
+        @test point ∈ box
+        key_2 = GAIO.point_to_key(partition, box.center)
```
Comment (cafaxo) [inline test/partition_regular.jl:72]: underscore looks weird

**Comment by cafaxo** [comment]
Some style issues still, but much better than before!
Nice job!

File: test/boxmap.jl (hunk 10 above)
Comment (MaHermann) [reply test/boxmap.jl:53]: I like to throw things like this in from time to time, because, almost by definition, tests tend to fail in unexpected ways. The basic idea was that with something like this, we can make sure that the error we test with `@test_throws` in the next line doesn't stem from some mistake in the definition of x or f, but actually from g, as it should. But it it also partly a leftover from before the refactoring today, and I agree that it probably is more confusing than helpful, so will remove it.

File: test/boxset.jl (hunk 12 above)
Comment (MaHermann) [reply test/boxset.jl:104]: The idea here and in the following was that these points are closely tied to the testset and are only outside as we reuse them in later testsets, so I didn't want to separate them to much from the testsets.

File: test/partition_tree.jl (hunk 14 above)
Comment (MaHermann) [reply test/partition_tree.jl:3]: I like to have this sort of things in place as a reminder and mild warning (a sort of "weaker TODO"). But I agree, for a public library this is probably not the right way.

**Comment by MaHermann** [comment]
Thanks, and thank you for your comments! To me, Julia feels a bit like trying to speak Spanish after having French in high school, so I'm still lacking a feeling for idiomatics. I should probably spend some time reading style guides :)

**Comment by MaHermann** [comment]
By the way, I plan on implementing a reduced version of the examples as you suggested, but probably won't mange to this week. So I will postpone that to a later PR.

File: test/partition_regular.jl [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+            partition = subdivide(partition)
+        end
+        @test depth(partition) == n
+        @test dimension(partition) == 4
+    end
+    @testset "size" begin
+        partition = RegularPartition(Box(SVector(0.0, 1.0), SVector(1.0, 1.0)), 3)
+        @test depth(partition) == 3
+        @test size(partition) == (4, 2)
+        @test prod(Base.size(partition)) == 2^depth(partition)
```
Comment (cafaxo) [inline test/partition_regular.jl:24]: No need for base prefix

File: test/boxmap.jl [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+        mapped1 = g(boxset)
+        mapped2 = g(boxset; target=boxset)
+        mapped3 = g(boxset; target=image)
+        mapped4 = g(boxset, target=full_boxset)
+
+        @test !isempty(boxset) && !isempty(image)
+        @test !isempty(mapped1) && !isempty(mapped2)
+        @test !isempty(mapped3) && !isempty(mapped4)
+        # easiest way right now to check for equality
+        @test length(union(image, mapped1)) == Base.length(intersect(image, mapped1))
```
Comment (cafaxo) [inline test/boxmap.jl:38]: you can remove all Base.
use find & replace :)
